<h1><center>Social Media Sentiment Analysis</center></h1>.

**Analyze social media data (e.g., Twitter) to understand public sentiment
towards specific topics, products, or events.\
 Use natural languag 
processing (NLP) techniques to preprocess text data, extract sentime t
scores, and visualize sentiment trends over ti**me.

## Import Libraries

In [161]:

import pandas as pd
import numpy as np
import nltk
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from nltk.stem.porter import PorterStemmer
import re 
from sklearn.model_selection import train_test_split


## Load Data

In [162]:
twitter_data = pd.read_csv('twitter.csv')
twitter_data.head()

,id,label,tweet
0,1,0,@user when a father is dysfunctional and is s...
1,2,0,@user @user thanks for #lyft credit i can't us...
2,3,0,bihday your majesty
3,4,0,#model i love u take with u all the time in ...
4,5,0,factsguide: society now #motivation


## Exploratory Data Analysis

In [163]:
twitter_data.shape

(31962, 3)

In [164]:
twitter_data['label'].value_counts()

label
0    29720
1     2242
Name: count, dtype: int64

In [165]:
twitter_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31962 entries, 0 to 31961
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      31962 non-null  int64 
 1   label   31962 non-null  int64 
 2   tweet   31962 non-null  object
dtypes: int64(2), object(1)
memory usage: 749.2+ KB


## Preprocess The Data

In [166]:
from nltk.stem import PorterStemmer
port_stem = PorterStemmer()


In [167]:
from nltk.corpus import stopwords
import nltk
nltk.download('stopwords')  # Download stopwords if not already done


[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\anamn\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

## **Stemming**
#### *Stemming is the process of reducing a word to its Root Word*
**example: actor, acting, actress = act**

In [168]:
# Initialize the stemmer
port_stem = PorterStemmer()

def stemming(content):
    # Remove non-alphabet characters
    stemmed_content = re.sub('[^a-zA-Z]', ' ', content)
    # Convert to lowercase
    stemmed_content = stemmed_content.lower()
    # Split into words
    stemmed_content = stemmed_content.split()
    # Apply stemming and remove stopwords
    stemmed_content = [port_stem.stem(word) for word in stemmed_content if word not in stopwords.words('english')]
    # Join back into a single string
    stemmed_content = ' '.join(stemmed_content)
    return stemmed_content


In [169]:
# Apply the function to the DataFrame
twitter_data['stemmed_content'] = twitter_data['tweet'].apply(stemming)

In [170]:
twitter_data.head()

,id,label,tweet,stemmed_content
0,1,0,@user when a father is dysfunctional and is s...,user father dysfunct selfish drag kid dysfunct...
1,2,0,@user @user thanks for #lyft credit i can't us...,user user thank lyft credit use caus offer whe...
2,3,0,bihday your majesty,bihday majesti
3,4,0,#model i love u take with u all the time in ...,model love u take u time ur
4,5,0,factsguide: society now #motivation,factsguid societi motiv


In [171]:
print(twitter_data['stemmed_content'])

0        user father dysfunct selfish drag kid dysfunct...
1        user user thank lyft credit use caus offer whe...
2                                           bihday majesti
3                              model love u take u time ur
4                                  factsguid societi motiv
                               ...                        
31957                                   ate user isz youuu
31958    see nina turner airwav tri wrap mantl genuin h...
31959             listen sad song monday morn otw work sad
31960     user sikh templ vandalis calgari wso condemn act
31961                                    thank user follow
Name: stemmed_content, Length: 31962, dtype: object


In [172]:
print(twitter_data['label'])

0        0
1        0
2        0
3        0
4        0
        ..
31957    0
31958    0
31959    0
31960    1
31961    0
Name: label, Length: 31962, dtype: int64


### Separating the data and label

In [173]:
X = twitter_data['stemmed_content'].values
y = twitter_data['label'].values

In [174]:
print(X)

['user father dysfunct selfish drag kid dysfunct run'
 'user user thank lyft credit use caus offer wheelchair van pdx disapoint getthank'
 'bihday majesti' ... 'listen sad song monday morn otw work sad'
 'user sikh templ vandalis calgari wso condemn act' 'thank user follow']


In [175]:
print(y)

[0 0 0 ... 0 1 0]


### Splitting the data to training and test data

In [176]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, stratify= y, random_state=2)
# stratify, in order to tell train_test split we want equal propotion of the two classes which is '0'& '1'both in training & test data.

In [177]:
print(X.shape, X_train.shape, X_test.shape)

(31962,) (25569,) (6393,)


In [178]:
print(X_train)

['user user user rightli gop hate trump obama'
 'user gtfoh itz crackalog malign black woman abt rais son instead arrest white man choke'
 'user true' ... 'messag make'
 'jokeoftheday pun tornado black peopl common take ruin neighborhood'
 'morn offic routin never without bff imi']


In [179]:
print(X_test)

['life way mum two fiona fraser pass away suddenli sunday morn'
 'next book read thefateoften iamnumberfourseri pittacuslor mamayakasakin'
 'think anyth wors date guy interest footbal flowerman eurosbitch' ...
 'tomorrow ethiad waitedyear'
 'new logo new product updat new product soon staup' 'want wake alex arm']


## Feature Extraction
**Converting the Textual data to Numerical data**

In [180]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Initialize the vectorizer
vectorizer = TfidfVectorizer()

# Fit the vectorizer on the training data and transform it
X_train = vectorizer.fit_transform(X_train)

# Transform the test data using the same vectorizer
X_test = vectorizer.transform(X_test)


In [181]:
print(X_train)

  (0, 16993)	0.3716656071840357
  (0, 24664)	0.3109447242132589
  (0, 10288)	0.32990469051375404
  (0, 9610)	0.4182753064960481
  (0, 20062)	0.5974288850441769
  (0, 25239)	0.3527966062591804
  (1, 4298)	0.30694007737757334
  (1, 14495)	0.1827892977387553
  (1, 26231)	0.18906535223896268
  (1, 1182)	0.27619659335065816
  (1, 11764)	0.23846443766142886
  (1, 22069)	0.2150334376242663
  (1, 19392)	0.2551565754522681
  (1, 81)	0.2639865531842322
  (1, 26513)	0.2118661382456014
  (1, 2572)	0.18698883185394868
  (1, 14469)	0.3141196985545238
  (1, 5199)	0.33642163892870125
  (1, 12042)	0.33642163892870125
  (1, 9919)	0.33642163892870125
  (1, 25239)	0.06622177995150041
  (2, 24653)	0.949186511103138
  (2, 25239)	0.3147141038146409
  (3, 8163)	0.30581104937005266
  (3, 18816)	0.42145904832974024
  :	:
  (25564, 15204)	0.28067049669302796
  (25564, 3790)	0.3447321388169406
  (25564, 24664)	0.22225173407837054
  (25564, 25239)	0.08405530137921485
  (25565, 24153)	0.399915584259267
  (25565, 18

In [182]:
print(X_test)

  (0, 25881)	0.26039296678434926
  (0, 24851)	0.3214533525354732
  (0, 22964)	0.28557788315916544
  (0, 22866)	0.47380967427831155
  (0, 17765)	0.35987967287211464
  (0, 15745)	0.408273860899287
  (0, 15534)	0.2705302892461171
  (0, 13526)	0.2243595720096581
  (0, 1462)	0.32107340408614415
  (1, 19530)	0.6038929722997515
  (1, 16423)	0.5459330064148513
  (1, 2887)	0.5807498863657343
  (2, 26660)	0.4452512256194358
  (2, 23985)	0.2883796021561549
  (2, 11795)	0.41139366639087444
  (2, 10002)	0.3154815775649124
  (2, 8461)	0.395007832073356
  (2, 5691)	0.3870776512436752
  (2, 1000)	0.37941386522327236
  (3, 22199)	0.47012536178994263
  (3, 12251)	0.5062853891596106
  (3, 9637)	0.6040463494357774
  (3, 5535)	0.3182839143772333
  (3, 788)	0.23765648846053805
  (4, 24380)	0.3631798301995126
  :	:
  (6386, 13439)	0.3586509065661895
  (6386, 9162)	0.17123123349495378
  (6386, 2509)	0.27317402052078327
  (6386, 2020)	0.27534016944483114
  (6386, 92)	0.38219998902009766
  (6387, 20882)	0.58582

### Training the 'Machine Learning Model'
**Logistic Regression**\
It is Classification Model to classify the data points into different classes

In [183]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(max_iter=1000)

In [184]:
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

### Model Evaluation
**Accuracy Score**

In [185]:
from sklearn.metrics import accuracy_score
# Accuracy score on the training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(y_train, X_train_prediction)*100
# we are comparing the true values(y_train)and the values predicted by our model

In [186]:
print('Accuracy_score on the training data :', training_data_accuracy)

Accuracy_score on the training data : 95.40068051155697


In [187]:

# Accuracy score on the test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(y_test, X_test_prediction)*100

In [188]:
print('Accuracy_score on the test data :', test_data_accuracy)

Accuracy_score on the test data : 95.11966213045518


**Logistic Regression model demonstrates excellent performance, with 95.12% test accuracy and 95.40% training accuracy.** \
**The small difference between the two indicates no significant overfitting, meaning the model generalizes well to new data. This high accuracy suggests the model effectively captures patterns in the dataset.**

**Model Accuracy = 95.11%**

In [189]:
from sklearn.metrics import classification_report

# Get detailed classification report
print(classification_report(y_test, X_test_prediction))


              precision    recall  f1-score   support

           0       0.95      1.00      0.97      5945
           1       0.91      0.33      0.49       448

    accuracy                           0.95      6393
   macro avg       0.93      0.67      0.73      6393
weighted avg       0.95      0.95      0.94      6393



#### **-Class 0 (negative sentiment):** The model performs exceptionally well, with 95% precision and 100% recall, meaning it correctly identifies almost all negative sentiments.

#### **-Class 1 (positive sentiment):** The model has high precision (91%), but low recall (33%), meaning it misses many actual positive sentiment tweets (high false negatives).

In [190]:
import numpy as np 
fnl = pd.DataFrame(np.c_[y_test, X_test_prediction], columns = ['Actual','Predicted'])
fnl

,Actual,Predicted
0,0,0
1,0,0
2,0,0
3,0,0
4,0,0
...,...,...
6388,0,0
6389,0,0
6390,0,0
6391,0,0


#### **np.c_[]:** This is a NumPy function that concatenates arrays along the second axis (columns). In this case, it combines the true labels (y_test) and the predicted labels (X_test_prediction) into a single 2D array.

#### **pd.DataFrame():** This converts the 2D array into a Pandas DataFrame, which is a tabular data structure that can hold different types of data.

#### **columns = ['Actual', 'Predicted']:** This specifies the column names for the DataFrame. The first column will be labeled 'Actual' (representing the true labels), and the second column will be labeled 'Predicted' (representing the model's predictions).
#### **The resulting DataFrame fnl contains two columns:** one for the actual sentiment labels from the test set and another for the predicted sentiment labels, allowing for easy comparison of model performance.**

## Summary:
**The model excels at identifying negative sentiments (Class 0), achieving 95% precision and 100% recall. \
However, it struggles with positive sentiments (Class 1), with high precision (91%) but low recall (33%), indicating many missed actual positive tweets.**